src/shogun/lib/SGMatrix.h

```diff
@@ -41,6 +41,49 @@ template<class T> class SGMatrix : public SGReferencedData
 	public:
 		typedef RandomIterator<T> iterator;
 		typedef ConstRandomIterator<T> const_iterator;
+	
+		class column_iterator
+		{
+		public:
+			//Iterator traits
+			using difference_type = index_t;
+			using value_type = SGVector<T>;
+			using pointer = SGVector<T>*;
+			using reference = SGVector<T>&;
+			using iterator_category = std::random_access_iterator_tag;
+
+			column_iterator(SGMatrix<T>& mat, index_t idx = 0) : m_mat(mat), m_idx(idx) {}
+			column_iterator(const column_iterator& other) : m_mat(other.m_mat), m_idx(other.m_idx) {}
+			
+			column_iterator& operator++(){ ++m_idx;return *this; }
+			column_iterator operator++(int) { column_iterator retval(*this); ++(*this); return retval;}
+			column_iterator& operator--(){ --m_idx;return *this; }
+			column_iterator operator--(int) { column_iterator retval(*this); --(*this); return retval;}
+
+			bool operator==(const column_iterator& other) const { return m_mat == other.m_mat && m_idx == other.m_idx; }
+			bool operator!=(const column_iterator& other) const { return !(*this == other); }
+
+			value_type operator*(){	return m_mat.get_column(m_idx); }
+			
+			column_iterator &operator += (difference_type d) { m_idx += d; return *this; }
+			column_iterator &operator -= (difference_type d) { m_idx -= d; return *this; }
+
+			column_iterator operator + (difference_type d) const { return column_iterator(m_mat,m_idx+d); }
+			column_iterator operator - (difference_type d) const { return column_iterator(m_mat,m_idx-d); }
+
+			value_type operator [] (difference_type d) const { return *column_iterator(m_mat,m_idx+d); }
+
+			bool operator < (const column_iterator &other) const { return m_idx < other.m_idx; }
+			bool operator > (const column_iterator &other) const { return m_idx > other.m_idx; }
+			bool operator <= (const column_iterator &other) const { return m_idx <= other.m_idx; }
+			bool operator >= (const column_iterator &other) const { return m_idx >= other.m_idx; }
+
+			difference_type operator - (const column_iterator &other) const {return m_idx - other.m_idx;}
+
+		private:
+			SGMatrix<T>& m_mat;
+			index_t m_idx;
```

Comment:
Maybe call this m_column_idx, so it is clear what it is tracking

Reply:
Sure.

Reply:
How about ```m_col_idx``` ?

Reply:
Sure!